### DISCUSSION TITLE
Folium integration: How to create a custom element?

### DISCUSSION THREAD

**Discussion by chr1st1ank** [zauberzeug/nicegui#487] [Q&A] [answered]
Hi all,

how can one create a custom element consisting of multiple HTML entities? Ideally without writing a custom Javascript component.

What I want to do is to add a [Folium](https://python-visualization.github.io/folium/index.html) map to a page. I have a nice visualization implemented in Folium with different markers with custom colors and text, all without writing javascript. It works like a charm in a Jupyter Notebook and I want to display it in the NiceGUI app.
I looked at the [leaflet example](https://github.com/zauberzeug/nicegui/tree/main/examples/map/) and tried to re-implement it from scratch without Folium, but find it really hard. 
It would be better to use Folium which wraps leaflet and offers a Python API.

What I can get from Folium are three parts that need to be added to the HTML:
- A header section with some script includes and a style tag
- A bit of generated javascript in a `<script>` tag that needs to be put somewhere into the HTML body
- A `div` tag as placeholder for the map which needs to be put into the right place in the HTML body. So e.g. inside a NiceGUI "card" element. It looks like this: `<div class="folium-map" id="map_664d30c7d59f99316a7b5d0243700ee8" ></div>`

Thanks for any good hint, e.g. to an example I can start from.

**Comment by AL-THE-BOT-FATHER** [comment]
can you post the code?

**Reply by chr1st1ank** [reply]
Thanks for following up. Not sure which of my different failing attempts is a good basis. My own starting point is this approach suggested in the [Folium documentation](https://python-visualization.github.io/folium/flask.html):

```python
from flask import Flask, render_template_string
import folium

app = Flask(__name__)

@app.route("/components")
def components():
    """Extract map components and put those on a page."""
    m = folium.Map(
        width=800,
        height=600,
    )

    m.get_root().render()
    header = m.get_root().header.render()
    body_html = m.get_root().html.render()
    script = m.get_root().script.render()

    return render_template_string(
        """
            <!DOCTYPE html>
            <html>
                <head>
                    {{ header|safe }}
                </head>
                <body>
                    <h1>Using components</h1>
                    {{ body_html|safe }}
                    <script>
                        {{ script|safe }}
                    </script>
                </body>
            </html>
        """,
        header=header,
        body_html=body_html,
        script=script,
    )
```

Here `header` is something like this:
<details>

    <meta http-equiv="content-type" content="text/html; charset=UTF-8" />
    <script>
        L_NO_TOUCH = false;
        L_DISABLE_3D = false;
    </script>
    <style>
        html,
        body {
            width: 100%;
            height: 100%;
            margin: 0;
            padding: 0;
        }
    </style>
    <style>
        #map {
            position: absolute;
            top: 0;
            bottom: 0;
            right: 0;
            left: 0;
        }
    </style>
    <script src="https://cdn.jsdelivr.net/npm/leaflet@1.9.3/dist/leaflet.js"></script>
    <script src="https://code.jquery.com/jquery-1.12.4.min.js"></script>
    <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.2.2/dist/js/bootstrap.bundle.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/Leaflet.awesome-markers/2.0.2/leaflet.awesome-markers.js"></script>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/leaflet@1.9.3/dist/leaflet.css" /> <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@5.2.2/dist/css/bootstrap.min.css" />
    <link rel="stylesheet" href="https://netdna.bootstrapcdn.com/bootstrap/3.0.0/css/bootstrap.min.css" /> <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@6.2.0/css/all.min.css" />
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/Leaflet.awesome-markers/2.0.2/leaflet.awesome-markers.css" />
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/python-visualization/folium/folium/templates/leaflet.awesome.rotate.min.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no" />
    <style>
        #map_664d30c7d59f99316a7b5d0243700ee8 {
            position: relative;
            width: 1000px;
            height: 500px;
            left: 0%;
            top: 0%;
        }
        .leaflet-container {
            font-size: 1rem;
        }
    </style>

</details>

`body_html` is simply `<div class="folium-map" id="map_664d30c7d59f99316a7b5d0243700ee8" ></div>`

The actual content for the map is defined in the script element which goes into the body:

<details>

    <script>
    var map_5a3ae88472c4c6e18093d06a5e96bef6 = L.map("map_5a3ae88472c4c6e18093d06a5e96bef6", {
        center: [0.0, 0.0],
        crs: L.CRS.EPSG3857,
        zoom: 10,
        zoomControl: true,
        preferCanvas: false,
    });
    
    var tile_layer_a7287746d65975b3addb95c2bcb99210 = L.tileLayer("https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png", {
        attribution: 'Data by \u0026copy; \u003ca target="_blank" href="http://openstreetmap.org"\u003eOpenStreetMap\u003c/a\u003e, under \u003ca target="_blank" href="http://www.openstreetmap.org/copyright"\u003eODbL\u003c/a\u003e.',
        detectRetina: false,
        maxNativeZoom: 18,
        maxZoom: 18,
        minZoom: 0,
        noWrap: false,
        opacity: 1,
        subdomains: "abc",
        tms: false,
    }).addTo(map_5a3ae88472c4c6e18093d06a5e96bef6);
    
    var marker_b74f238c4cb19c3f2e272c69b708586f = L.marker([3.0337019, 101.7653809], {}).addTo(map_5a3ae88472c4c6e18093d06a5e96bef6);
    
    marker_b74f238c4cb19c3f2e272c69b708586f.bindTooltip(
        `<div><p>Nice tooltip</p></div>`,
        { sticky: true }
    );
    
    var marker_18ec9bd7e47409e07706ea732c4429f2 = L.marker([3.0337019, 101.7653809], {}).addTo(map_5a3ae88472c4c6e18093d06a5e96bef6);
    
    var icon_3aa81ffbe84238b0494b0742be400f24 = L.AwesomeMarkers.icon({ extraClasses: "fa-rotate-0", icon: "info-sign", iconColor: "white", markerColor: "red", prefix: "glyphicon" });
    marker_18ec9bd7e47409e07706ea732c4429f2.setIcon(icon_3aa81ffbe84238b0494b0742be400f24);
    
    marker_18ec9bd7e47409e07706ea732c4429f2.bindTooltip(
        `<div><p>Another nice tooltip</p></div>`,
        { sticky: true }
    );
    
    map_5a3ae88472c4c6e18093d06a5e96bef6.fitBounds(
        [
            [3.0337019, 101.7653809],
            [3.0337019, 101.7653809],
        ],
        {}
    );
    
    var lat_lng_popup_3ac1b99f592b44deccc952d5c1e8467d = L.popup();
    function latLngPop(e) {
        lat_lng_popup_3ac1b99f592b44deccc952d5c1e8467d
            .setLatLng(e.latlng)
            .setContent("Latitude: " + e.latlng.lat.toFixed(4) + "<br>Longitude: " + e.latlng.lng.toFixed(4))
            .openOn(map_5a3ae88472c4c6e18093d06a5e96bef6);
    }
    map_5a3ae88472c4c6e18093d06a5e96bef6.on("click", latLngPop);
    </script>

</details>

**Comment by chr1st1ank** [comment]
Ok, now I have an implementation which could almost work. It looks like this:

```python
import folium
from nicegui import ui
from nicegui.element import Element

class Map(Element):
    def __init__(self, folium_map: folium.Map) -> None:
        super().__init__(tag="div")
        self._classes = ["w-full"]

        self.folium_map = folium_map
        self.folium_map.get_root().render()

        ui.add_head_html(self.folium_map.get_root().header.render())
        ui.html(self.folium_map.get_root().html.render())
        ui.add_body_html(f"<script>{folium_map.get_root().script.render().strip()}</script>")
```

This gives the following error in the Browser's JS console:
```
Map.js:1092 
       Uncaught Error: Map container not found.
    at e._initContainer (Map.js:1092:10)
    at e.initialize (Map.js:136:8)
    at new e (Class.js:24:20)
    at t.map (Map.js:1750:9)
    at test:54:58
```

I think, the reason is, that the leaflet script in the html body tries to access the `<div>` element by ID before it is there.
So if there is any possibility in NiceGUI to send the result of `ui.html()` down to the browser and only when it is there send the script via `ui.add_body_html()` , it should work.

**Reply by chr1st1ank** [reply]
The script is there:
![image](https://user-images.githubusercontent.com/1138025/224181899-57f1be7f-4f5e-4813-9571-c544b82f2113.png)

and the div also:
![image](https://user-images.githubusercontent.com/1138025/224181963-357d992b-9e0e-42c2-8378-957ad0ca1540.png)

So, it must be the timing

**Comment by falkoschindler** [comment]
A NiceGUI component for Folium sounds like a great idea!

I think, however, that it will be pretty hard or even impossible to achieve without a custom Vue/JS component. As you notice, you need to control when to load what code and when to run what script. And there is more: How to update the client when the Folium map changes? How to deal with multiple maps on the same page? We don't want to load the same head HTML multiple times. And how to handle multiple clients on a shared page?

All these considerations lead to implementations like `aggrid.js`, `mermaid.js`, and `plotly.vue`. I can imagine that a Folium component might work in a very similar way.

**Reply by rodja** [reply]
Exactly my thoughts. Why would you like to specifically "not use JavaScript" to control the JavaScript part of this library?

**Reply by chr1st1ank** [reply]
I mean, yes, using the right tool for the job is necessary. I just would have hoped that nicegui provides an abstraction layer around the js.
Especially sophisticated things like custom components are not really accessible to a programmer with just basic JS experience like me.
But if it doesn't work out for Folium, JS it must be. I'm just asking myself if Folium is still the right tool then. Pulling the JS output of a python wrapper around JS and then using more JS to marry it with another Python wrapper around other JS sounds not like the task I've been dreaming of ever since 🥶

**Reply by chr1st1ank** [reply]
Another thought: maybe it doesn't work with Folium specifically. But the ability to add a custom element as Python class which adds some script and CSS imports to the header just once on a page and then adds a html element in the place in the DOM tree where the object is instantiated would make nicegui much easier extendable. Maybe that's possible and only needs to be documented.

**Reply by falkoschindler** [reply]
> I'm just asking myself if Folium is still the right tool then.

I think so, yes. Having an existing Python API simplifies the user code a lot compared to sending JavaScript commands back and forth to interact with leaflet. Sure, the interface between Python and JavaScript has to be written somewhere, and that can get tricky. But with AG Grid, for example, this works very well and NiceGUI users are building upon AG Grid features that we are not even aware of, simply because you can generically call any AG Grid API method.

> Pulling the JS output of a python wrapper around JS and then using more JS...

You're probably right: Extracting `header`, `html`, and `script` from Folium and somehow putting it back together in a custom JS component sounds weird, especially considering the contents of these strings. I would probably try to use `folium_map.get_root().to_json()` and pass this "raw" representation to leaflet. But I haven't looked deeper into it.

> But the ability to add a custom element as Python class...

That's an interesting idea. We'll think about it.

**Comment by chr1st1ank** [comment, marked as answer]
It turns out, that the simplest way to embed a folium map is, to use an iframe. A folium map has the method `_repr_html_()` which is also used in jupyter notebooks. It returns a self-contained `<iframe>` element which can be embedded with ui.html().

This isn't the silver bullet, e.g. it isn't possible to trigger any interaction from the map, but it is enough if the only goal is to display a custom scrollable and zoomable map. Working example:

```python
import folium
from nicegui import ui

with ui.card():
    map_width = 600
    map_height = 300
    m = folium.Map(location=[32.3, 295.21], width=map_width, height=map_height)
    m.get_root().width = f"{map_width}px"
    m.get_root().height = f"{map_height}px"
    iframe = m.get_root()._repr_html_()
    ui.html(iframe).classes("w-full h-full")
```

**Reply by falkoschindler** [reply]
Wow, thanks for this update! So it turns out you only need to create the Folium map and pass the HTML representation to `ui.html`. In a (tiny) nutshell:

```py
m = folium.Map()
ui.html(m.get_root()._repr_html_())
```

Of course, this approach is limited. But it might be enough in many cases. Awesome!
